ext/strings_test.go

```diff
@@ -329,3 +333,467 @@ func TestStringsWithExtension(t *testing.T) {
+			format:         "true bool: %s, false bool: %s",
+			formatArgs:     `true, false`,
+			expectedOutput: "true bool: true, false bool: false",
+		},
+		{
+			name:       "unrecognized formatting clause",
+			format:     "%a",
+			formatArgs: "1",
+			err:        "could not parse formatting clause: unrecognized formatting clause \"a\"",
+		},
+		{
+			name:       "out of bounds arg index",
+			format:     "%d %d %d",
+			formatArgs: "0, 1",
+			err:        "index 2 out of range",
+		},
+		{
+			name:       "string substitution is not allowed with binary clause",
+			format:     "string is %b",
+			formatArgs: `"abc"`,
+			err:        "error during formatting: only integers and bools can be formatted as binary, was given string",
+		},
+		{
+			name:       "int substitution not allowed with string clause",
+			format:     "int is %s",
+			formatArgs: "1",
+			err:        "error during formatting: string clause can only be used on strings, bools, bytes, maps, lists, durations, and timestamps, was given int",
+		},
+		{
+			name:       "duration substitution not allowed with decimal clause",
+			format:     "%d",
+			formatArgs: `duration("30m2s")`,
+			err:        "error during formatting: decimal clause can only be used on integers, was given google.protobuf.Duration",
+		},
+		{
+			name:       "double substitution not allowed with string clause",
+			format:     "double is %s",
+			formatArgs: "0.5",
+			err:        "error during formatting: string clause can only be used on strings, bools, bytes, maps, lists, durations, and timestamps, was given double",
+		},
+		{
+			name:       "string substitution not allowed with octal clause",
+			format:     "octal: %o",
+			formatArgs: `"a string"`,
+			err:        "error during formatting: octal clause can only be used on integers, was given string",
+		},
+		{
+			name:       "double substitution not allowed with hex clause",
+			format:     "double is %x",
+			formatArgs: "0.5",
+			err:        "error during formatting: only integers, byte buffers, and strings can be formatted as hex, was given double",
+		},
+		{
+			name:       "uppercase not allowed for scientific clause",
+			format:     "double is %E",
+			formatArgs: "0.5",
+			err:        `could not parse formatting clause: unrecognized formatting clause "E"`,
+		},
+		{
+			name:       "object not allowed",
+			format:     "object is %s",
+			formatArgs: `ext.TestAllTypes{PbVal: test.TestAllTypes{}}`,
+			err:        "error during formatting: string clause can only be used on strings, bools, bytes, maps, lists, durations, and timestamps, was given ext.TestAllTypes",
+		},
+		{
+			name:       "enum is treated as int",
+			format:     "enum is %s",
+			formatArgs: `ext.TestAllTypes{PbVal: test.TestAllTypes{standalone_enum: 1}}.PbVal.standalone_enum`,
+			err:        "error during formatting: string clause can only be used on strings, bools, bytes, maps, lists, durations, and timestamps, was given int",
+		},
```

Comment:
Apologies for noticing this so late in the review process, but I just realized we have no tests for `dyn` types. Would you please add? I'd like to be sure that if we don't have static type information that runtime types are still handled well.

Reply:
Gotcha, I've added an equivalent coverage for `dyn` where applicable.